Bug 1169320 - Add support to ingest jobs via Pulse

camd opened the pull request:
Addresses [Bug 1169320](https://bugzilla.mozilla.org/show_bug.cgi?id=1169320)

This branch enables ingesting jobs from a pulse exchange in the specified YAML format provided here.

Deferred to follow-up PRs:

* Update existing queues based on config changes
* API endpoints to expose the schema


<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/mozilla/treeherder/940)
<!-- Reviewable:end -->

camd commented on treeherder/etl/pulse_consumer.py:
```diff
@@ -0,0 +1,227 @@
+
+    # status can only transition from lower to higher order.  If a job comes
+    # in to update a status to a lower order, it will be skipped as out of
+    # sequence.
```
I need to validate this in ``jobs.py`` to ensure it works correctly.

camd commented on treeherder/etl/pulse_consumer.py:
```diff
@@ -0,0 +1,228 @@
+
+    @staticmethod
+    def get_jobs_schema():
+        with open("{0}/schemas/job.yml".format(
```
need a better way to load this file in.  Perhaps from a URL and cache it.

edmorley commented:
Review status: 0 of 12 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

<sup>**[requirements/common.txt, line 110 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/940#-JyvXNdLj7bY0q1QxAXo)** ([raw file](https://github.com/mozilla/treeherder/blob/f467ba07ac475a5edaf74f0b0f0b015598a726aa/requirements/common.txt#L110)):</sup>
python-dateutil requires the package six - please move six from dev.txt to common.txt and add a mentioning the dependency :-)

---

<sup>**[treeherder/model/sql/jobs.json, line 285 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/940#-JyvY-Nu6FuvdmxlDrKh)** ([raw file](https://github.com/mozilla/treeherder/blob/f467ba07ac475a5edaf74f0b0f0b015598a726aa/treeherder/model/sql/jobs.json#L285)):</sup>
It looks like this + the corresponding change in jobs.py is fixing a pre-existing issue? Could you split this out to another commit or another bug?

---

<sup>**[treeherder/settings/base.py, line 360 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/940#-JyvYI9uTnfjaZybu6Fl)** ([raw file](https://github.com/mozilla/treeherder/blob/f467ba07ac475a5edaf74f0b0f0b015598a726aa/treeherder/settings/base.py#L360)):</sup>
We're deprecating local.py - I think this would be better as a URL that we parse, like we have for DATABASE_URL etc. This can either be done manually, like on:
https://devcenter.heroku.com/articles/cloudamqp

Or else we can use django-environ's url() - which I was going to start us using in bug 1160561 anyway.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/940)
<!-- Sent from Reviewable.io -->

maurodoglio commented on treeherder/etl/management/commands/ingest_from_pulse.py:
```diff
@@ -0,0 +1,44 @@
+
+class Command(BaseCommand):
+
+    """Management command to publish a job to a pulse exchange."""
```
The docstrings and the help string disagree

maurodoglio commented on treeherder/etl/schemas/job.yml:
```diff
@@ -0,0 +1,278 @@
+    maximum: 3
+
+  isRetried:
+    description: True indicates this job is has been retried.
```
s/is//

maurodoglio commented on treeherder/etl/schemas/job.yml:
```diff
@@ -0,0 +1,278 @@
+        blob:
+          type: "string"
+          minLength: 1
+          maxLength: 4096
```
I don't think 4096 characters is enough for a log file. But maybe I'm missing something.

camd replied:
For log files, we just hold the reference to it, not the log itself.  But for the new error summary files coming in with auto-starring, it is possible for them to be quite large.  Currently, we have no limit on size for an artifact, and perhaps we shouldn't put a limit yet.

camd commented:
Review status: 0 of 19 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

<sup>**[requirements/common.txt, line 110 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/940#-JyvXNdLj7bY0q1QxAXo)** ([raw file](https://github.com/mozilla/treeherder/blob/f467ba07ac475a5edaf74f0b0f0b015598a726aa/requirements/common.txt#L110)):</sup>
Done.

---

<sup>**[schemas/pulse-job.yml, line 119 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/940#-JyO0KxWy-r5mzXyUJzO-r4-119)** ([raw file](https://github.com/mozilla/treeherder/blob/2d2fcfd7571ede74a65ff5db40c88cceba95398c/schemas/pulse-job.yml#L119)):</sup>
Done.

---

<sup>**[schemas/pulse-job.yml, line 247 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/940#-JyO0KxWy-r5mzXyUJzO-r4-247)** ([raw file](https://github.com/mozilla/treeherder/blob/2d2fcfd7571ede74a65ff5db40c88cceba95398c/schemas/pulse-job.yml#L247)):</sup>
I removed the limit

---

<sup>**[treeherder/model/sql/jobs.json, line 285 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/940#-JyvY-Nu6FuvdmxlDrKh)** ([raw file](https://github.com/mozilla/treeherder/blob/f467ba07ac475a5edaf74f0b0f0b015598a726aa/treeherder/model/sql/jobs.json#L285)):</sup>
Done.

---

<sup>**[treeherder/settings/base.py, line 332 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/940#-JyvYI9uTnfjaZybu6Fl)** ([raw file](https://github.com/mozilla/treeherder/blob/f467ba07ac475a5edaf74f0b0f0b015598a726aa/treeherder/settings/base.py#L332)):</sup>
As discussed on Vidyo, I implemented using django-environ in one of the commits and a separate PR.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/940)
<!-- Sent from Reviewable.io -->

wlach commented:
Review status: 0 of 19 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

<sup>**[treeherder/etl/management/commands/ingest_from_pulse.py, line 31 \[r6\]](https://reviewable.io:443/reviews/mozilla/treeherder/940#-JzlxFQ3zQgvSms8aInc)** ([raw file](https://github.com/mozilla/treeherder/blob/14e07c8dee6a9fa6157232b4f03865db040544a4/treeherder/etl/management/commands/ingest_from_pulse.py#L31)):</sup>
Don't we normally use `self.stdout.write` for debugging info on management commands?

---

<sup>**[treeherder/webapp/schemas/pulse_jobs.py, line 21 \[r6\]](https://reviewable.io:443/reviews/mozilla/treeherder/940#-JzlxMlTiIlwOXbKhwOE)** ([raw file](https://github.com/mozilla/treeherder/blob/14e07c8dee6a9fa6157232b4f03865db040544a4/treeherder/webapp/schemas/pulse_jobs.py#L21)):</sup>
Nit: unnecessary newline

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/940)
<!-- Sent from Reviewable.io -->

camd commented:
Review status: 0 of 16 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

<sup>**[treeherder/etl/management/commands/ingest_from_pulse.py, line 30 \[r6\]](https://reviewable.io:443/reviews/mozilla/treeherder/940#-JzlxFQ3zQgvSms8aInc)** ([raw file](https://github.com/mozilla/treeherder/blob/14e07c8dee6a9fa6157232b4f03865db040544a4/treeherder/etl/management/commands/ingest_from_pulse.py#L30)):</sup>
Yeah, I think you're right.  Fixed.  

---

<sup>**[treeherder/webapp/schemas/pulse_jobs.py, line 0 \[r6\]](https://reviewable.io:443/reviews/mozilla/treeherder/940#-JzlxMlTiIlwOXbKhwOE)** ([raw file](https://github.com/mozilla/treeherder/blob/14e07c8dee6a9fa6157232b4f03865db040544a4/treeherder/webapp/schemas/pulse_jobs.py#L0)):</sup>
I deferred the APIs till a later, possible, PR.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/940)
<!-- Sent from Reviewable.io -->

gregarndt commented on schemas/pulse-job.yml:
```diff
@@ -0,0 +1,277 @@
+        properties:
+          kind:
+            type: "string"
+            enum: ['hg.mozilla.org']
```
Hrm, so it looks like origin could either be hg.mozilla.org or github....why the two different objects in the schema rather than list them both in the enum?

gregarndt commented on schemas/pulse-job.yml:
```diff
@@ -0,0 +1,277 @@
+          kind:
+            type: "string"
+            enum: ['github.com']
+          project:
```
So I assume that "project" is the same thing as the repository.  Is there a place to define what user that project/repo is for?

gregarndt commented on schemas/pulse-job.yml:
```diff
@@ -0,0 +1,277 @@
+        pattern: "^[A-Za-z0-9_-]+$"
+        minLength: 1
+        maxLength: 25
+      groupSymbol:
```
Currently we send in '?' for no group symbol...would that not be allowed anymore for this?  What should we send for group symbol if there is none (a 'B' that appears outside of a group)

maurodoglio commented:
Reviewed 11 of 13 files at r1, 5 of 5 files at r2, 5 of 6 files at r3, 9 of 9 files at r4, 17 of 17 files at r5, 1 of 1 files at r6, 17 of 17 files at r7, 6 of 6 files at r8, 3 of 3 files at r9, 2 of 2 files at r10.
Review status: all files reviewed at latest revision, 18 unresolved discussions, all commit checks successful.

---

<sup>**[treeherder/etl/pulse_consumer.py, line 48 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/940#-K-OtcMnPVXOX2BS-kyK)** ([raw file](https://github.com/mozilla/treeherder/blob/4b4df2f5c4dd99658fa8a226d67cd90684dc9e54/treeherder/etl/pulse_consumer.py#L48)):</sup>
shouldn't this be durable=True and auto_delete=False? I'm thinking of a case where the consumer gets restarted and we don't want to loose any pending job 
<!-- Reviewable comment -K-OtcMo_VE7B2tv_k9k:3222743 -->

---

<sup>**[treeherder/etl/pulse_consumer.py, line 63 \[r5\]](https://reviewable.io:443/reviews/mozilla/treeherder/940#-K-Ozr2NL9Cr_5AouvAI)** ([raw file](https://github.com/mozilla/treeherder/blob/8430b2e09942b5b8bc04f980376ffb4a557bbab0/treeherder/etl/pulse_consumer.py#L63)):</sup>
is the disconnect method called in case of a failure as well?
<!-- Reviewable comment -K-Ozr2NL9Cr_5AouvAJ:-626986332 -->

---

<sup>**[treeherder/etl/pulse_consumer.py, line 110 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/940#-K-OuIBu-6-QjuYpUIeC)** ([raw file](https://github.com/mozilla/treeherder/blob/4b4df2f5c4dd99658fa8a226d67cd90684dc9e54/treeherder/etl/pulse_consumer.py#L110)):</sup>
I guess this is a debug leftover. If I'm wrong, please log it with either debug or info level
<!-- Reviewable comment -K-OuIBu-6-QjuYpUIeD:836155359 -->

---

<sup>**[treeherder/etl/pulse_consumer.py, line 198 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/940#-K-Ov3Udfr4kNh4BYkuf)** ([raw file](https://github.com/mozilla/treeherder/blob/4b4df2f5c4dd99658fa8a226d67cd90684dc9e54/treeherder/etl/pulse_consumer.py#L198)):</sup>
Can we remove `unscheduled` from the list of accepted states? I guess when/if we need it, there will be a PR taking care of that with a change to the schema as well.
At the moment if you submit a job with `unscheduled` state it raises an error even if the schema was valid.
<!-- Reviewable comment -K-Ov3Udfr4kNh4BYkug:-1372138431 -->

---

<sup>**[tests/etl/test_pulse_consumer.py, line 20 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/940#-K-OwHB8iPZeUPMP20bi)** ([raw file](https://github.com/mozilla/treeherder/blob/4b4df2f5c4dd99658fa8a226d67cd90684dc9e54/tests/etl/test_pulse_consumer.py#L20)):</sup>
leftover pprint
<!-- Reviewable comment -K-OwHB8iPZeUPMP20bj:895537917 -->

---

<sup>**[treeherder/model/fixtures/repository.json, line 29 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/940#-K-OxCZNrY5SpGzN2x9h)** ([raw file](https://github.com/mozilla/treeherder/blob/0c2ece75d4ad7aae9a43c8089f812346e0a7fb4b/treeherder/model/fixtures/repository.json#L29)):</sup>
I don't know what happened, it looks like you deleted a lot of content from this file in the previous commit and you put it back in this one. Is it intentional?
<!-- Reviewable comment -K-OxCZNrY5SpGzN2x9i:-460840867 -->

---

<sup>**[treeherder/model/fixtures/repository.json, line 51 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/940#-K-OwVs9K0tK7n5ZaTIB)** ([raw file](https://github.com/mozilla/treeherder/blob/525866a553c95191a23168af7751d35bfc6c35ce/treeherder/model/fixtures/repository.json#L51)):</sup>
what is this change about?
<!-- Reviewable comment -K-OwVs9K0tK7n5ZaTIC:-844372513 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/940)
<!-- Sent from Reviewable.io -->

camd commented:
Review status: 12 of 16 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

<sup>**[schemas/pulse-job.yml, line 22 \[r10\]](https://reviewable.io:443/reviews/mozilla/treeherder/940#-JzWGKBU1C5VV-_ojkMn-r10-22)** ([raw file](https://github.com/mozilla/treeherder/blob/ef841906f271cd1915004f1e64f3397541dff6d1/schemas/pulse-job.yml#L22)):</sup>
Yeah, Jonas and I chatted about this a while.  This was so we could have a more descriptive field for either ``pushLogID`` or ``pullRequestID``
<!-- Reviewable comment -K-QBKPxHhXum7AZLJNU:-272321656 -->

---

<sup>**[schemas/pulse-job.yml, line 41 \[r10\]](https://reviewable.io:443/reviews/mozilla/treeherder/940#-JzWGKBU1C5VV-_ojkMn-r10-41)** ([raw file](https://github.com/mozilla/treeherder/blob/ef841906f271cd1915004f1e64f3397541dff6d1/schemas/pulse-job.yml#L41)):</sup>
Correct about product/repo.  But this is another good catch here.  The code had support for "productName", but I had removed it from the YML.  Fixed.
<!-- Reviewable comment -K-QB5jmSW_y6dz9Ftly:707902268 -->

---

<sup>**[schemas/pulse-job.yml, line 63 \[r10\]](https://reviewable.io:443/reviews/mozilla/treeherder/940#-JzWGKBU1C5VV-_ojkMn-r10-63)** ([raw file](https://github.com/mozilla/treeherder/blob/ef841906f271cd1915004f1e64f3397541dff6d1/schemas/pulse-job.yml#L63)):</sup>
Thanks for catching this.  I updated the YML and added some tests for several symbol combinations for both jobs and groups.
<!-- Reviewable comment -K-Q9tl5iIq1pOWMyvF8:-408297327 -->

---

<sup>**[treeherder/etl/pulse_consumer.py, line 0 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/940#-K-OtcMnPVXOX2BS-kyK)** ([raw file](https://github.com/mozilla/treeherder/blob/4b4df2f5c4dd99658fa8a226d67cd90684dc9e54/treeherder/etl/pulse_consumer.py#L0)):</sup>
Yeah, I think I agree.  I had been defaulting to the "ingesting locally" case.  But it's probably wiser to default to the deployment case and let the user override those values locally to make them transient, etc.   However I had neglected to allow them to specify ``auto_delete``, which was an error.  Thanks for catching that!
<!-- Reviewable comment -K-PB2nL1YpxQtBuNes_:2076932307 -->

---

<sup>**[treeherder/etl/pulse_consumer.py, line 63 \[r5\]](https://reviewable.io:443/reviews/mozilla/treeherder/940#-K-Ozr2NL9Cr_5AouvAI)** ([raw file](https://github.com/mozilla/treeherder/blob/8430b2e09942b5b8bc04f980376ffb4a557bbab0/treeherder/etl/pulse_consumer.py#L63)):</sup>
Yes.  It is called in a ``finally`` clause in the ``ingest_from_pulse`` management command.
<!-- Reviewable comment -K-PCZNNGz5B7-lAa-RF:1379368741 -->

---

<sup>**[treeherder/etl/pulse_consumer.py, line 111 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/940#-K-OuIBu-6-QjuYpUIeC)** ([raw file](https://github.com/mozilla/treeherder/blob/4b4df2f5c4dd99658fa8a226d67cd90684dc9e54/treeherder/etl/pulse_consumer.py#L111)):</sup>
It's hard for me to tell which line your comment refers to.  I think reviewable.io is kinda confused here.  If the line was ``logger.warn`` then it's not there anymore.  The current line is after an exception, it does a ``logger.error`` with the exception.
<!-- Reviewable comment -K-PD3mSkNfPZ_OUAOJ4:1783329840 -->

---

<sup>**[treeherder/model/fixtures/repository.json, line 29 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/940#-K-OxCZNrY5SpGzN2x9h)** ([raw file](https://github.com/mozilla/treeherder/blob/0c2ece75d4ad7aae9a43c8089f812346e0a7fb4b/treeherder/model/fixtures/repository.json#L29)):</sup>
it was just something I did while testing.  I had accidentally checked it in at one point, but restored it later.  Trimming this file down just makes vagrant provision and data ingestion go much faster.
<!-- Reviewable comment -K-PeOAagC92Kv1bccOE:-1772280166 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/940)
<!-- Sent from Reviewable.io -->
